File: android/app/src/main/java/com/xax/CryptoSavingsTracker/domain/usecase/execution/MonthlyCycleActionGate.kt
package com.xax.CryptoSavingsTracker.domain.usecase.execution

import com.xax.CryptoSavingsTracker.domain.model.UiCycleState

enum class MonthlyCycleAction {
    START_TRACKING,
    FINISH_MONTH,
    UNDO_START,
    UNDO_COMPLETION
}

enum class MonthlyCycleBlockedCopyKey {
    START_BLOCKED_ALREADY_EXECUTING,
    START_BLOCKED_CLOSED_MONTH,
    FINISH_BLOCKED_NO_EXECUTING,
    UNDO_START_EXPIRED,
    UNDO_COMPLETION_EXPIRED,
    RECORD_CONFLICT
}

data class MonthlyCycleActionDecision(
    val allowed: Boolean,
    val blockedCopyKey: MonthlyCycleBlockedCopyKey? = null,
    val blockedMessage: String? = null
)

object MonthlyCycleActionGate {
    fun evaluate(state: UiCycleState, action: MonthlyCycleAction): MonthlyCycleActionDecision {
        return when (state) {
            is UiCycleState.Planning -> {
                val month = state.monthLabel
                when (action) {
                    MonthlyCycleAction.START_TRACKING -> MonthlyCycleActionDecision(allowed = true)
                    MonthlyCycleAction.FINISH_MONTH -> blocked(
                        MonthlyCycleBlockedCopyKey.FINISH_BLOCKED_NO_EXECUTING,
                        ExecutionActionCopyCatalog.finishBlockedNoExecuting()
                    )
                    MonthlyCycleAction.UNDO_START -> blocked(
                        MonthlyCycleBlockedCopyKey.UNDO_START_EXPIRED,
                        ExecutionActionCopyCatalog.undoStartExpired(month)
                    )
                    MonthlyCycleAction.UNDO_COMPLETION -> blocked(
                        MonthlyCycleBlockedCopyKey.UNDO_COMPLETION_EXPIRED,
                        ExecutionActionCopyCatalog.undoCompletionExpired(month)
                    )
                }
            }

            is UiCycleState.Executing -> {
                val month = state.monthLabel
                when (action) {
                    MonthlyCycleAction.START_TRACKING -> blocked(
                        MonthlyCycleBlockedCopyKey.START_BLOCKED_ALREADY_EXECUTING,
                        ExecutionActionCopyCatalog.startBlockedAlreadyExecuting(month)
                    )
                    MonthlyCycleAction.FINISH_MONTH -> MonthlyCycleActionDecision(allowed = true)
                    MonthlyCycleAction.UNDO_START -> {
                        if (state.canUndoStart) {
                            MonthlyCycleActionDecision(allowed = true)
                        } else {
                            blocked(
                                MonthlyCycleBlockedCopyKey.UNDO_START_EXPIRED,
                                ExecutionActionCopyCatalog.undoStartExpired(month)
                            )
                        }
                    }
                    MonthlyCycleAction.UNDO_COMPLETION -> blocked(
                        MonthlyCycleBlockedCopyKey.UNDO_COMPLETION_EXPIRED,
                        ExecutionActionCopyCatalog.undoCompletionExpired(month)
                    )
                }
            }

            is UiCycleState.Closed -> {
                val month = state.monthLabel
                when (action) {
                    MonthlyCycleAction.START_TRACKING -> blocked(
                        MonthlyCycleBlockedCopyKey.START_BLOCKED_CLOSED_MONTH,
                        ExecutionActionCopyCatalog.startBlockedClosedMonth()
                    )
                    MonthlyCycleAction.FINISH_MONTH -> blocked(
                        MonthlyCycleBlockedCopyKey.FINISH_BLOCKED_NO_EXECUTING,
                        ExecutionActionCopyCatalog.finishBlockedNoExecuting()
                    )
                    MonthlyCycleAction.UNDO_START -> blocked(
                        MonthlyCycleBlockedCopyKey.UNDO_START_EXPIRED,
                        ExecutionActionCopyCatalog.undoStartExpired(month)
                    )
                    MonthlyCycleAction.UNDO_COMPLETION -> {
                        if (state.canUndoCompletion) {
                            MonthlyCycleActionDecision(allowed = true)
                        } else {
                            blocked(
                                MonthlyCycleBlockedCopyKey.UNDO_COMPLETION_EXPIRED,
                                ExecutionActionCopyCatalog.undoCompletionExpired(month)
                            )
                        }
                    }
                }
            }

            is UiCycleState.Conflict -> blocked(
                MonthlyCycleBlockedCopyKey.RECORD_CONFLICT,
                ExecutionActionCopyCatalog.recordConflict()
            )
        }
    }

    private fun blocked(
        key: MonthlyCycleBlockedCopyKey,
        message: String
    ): MonthlyCycleActionDecision {
        return MonthlyCycleActionDecision(
            allowed = false,
            blockedCopyKey = key,
            blockedMessage = message
        )
    }
}

File: ios/CryptoSavingsTracker/Services/MonthlyCycleActionGate.swift
//
//  MonthlyCycleActionGate.swift
//  CryptoSavingsTracker
//
//  Canonical action matrix: UiCycleState x Action -> allow/block + copy key.
//

import Foundation

enum MonthlyCycleAction {
    case startTracking
    case finishMonth
    case undoStart
    case undoCompletion
}

enum MonthlyCycleBlockedCopyKey {
    case startBlockedAlreadyExecuting
    case startBlockedClosedMonth
    case finishBlockedNoExecuting
    case undoStartExpired
    case undoCompletionExpired
    case recordConflict
}

struct MonthlyCycleActionDecision {
    let allowed: Bool
    let blockedCopyKey: MonthlyCycleBlockedCopyKey?
    let blockedMessage: String?

    static let allowedDecision = MonthlyCycleActionDecision(
        allowed: true,
        blockedCopyKey: nil,
        blockedMessage: nil
    )
}

enum MonthlyCycleActionGate {
    static func evaluate(state: UiCycleState, action: MonthlyCycleAction) -> MonthlyCycleActionDecision {
        switch state {
        case .planning(let month, _):
            switch action {
            case .startTracking:
                return .allowedDecision
            case .finishMonth:
                return blocked(.finishBlockedNoExecuting, MonthlyCycleCopyCatalog.finishBlockedNoExecuting())
            case .undoStart:
                return blocked(.undoStartExpired, MonthlyCycleCopyCatalog.undoStartExpired(month: month))
            case .undoCompletion:
                return blocked(.undoCompletionExpired, MonthlyCycleCopyCatalog.undoCompletionExpired(month: month))
            }

        case .executing(let month, _, let canUndoStart):
            switch action {
            case .startTracking:
                return blocked(.startBlockedAlreadyExecuting, MonthlyCycleCopyCatalog.startBlockedAlreadyExecuting(month: month))
            case .finishMonth:
                return .allowedDecision
            case .undoStart:
                return canUndoStart
                    ? .allowedDecision
                    : blocked(.undoStartExpired, MonthlyCycleCopyCatalog.undoStartExpired(month: month))
            case .undoCompletion:
                return blocked(.undoCompletionExpired, MonthlyCycleCopyCatalog.undoCompletionExpired(month: month))
            }

        case .closed(let month, let canUndoCompletion):
            switch action {
            case .startTracking:
                return blocked(.startBlockedClosedMonth, MonthlyCycleCopyCatalog.startBlockedClosedMonth())
            case .finishMonth:
                return blocked(.finishBlockedNoExecuting, MonthlyCycleCopyCatalog.finishBlockedNoExecuting())
            case .undoStart:
                return blocked(.undoStartExpired, MonthlyCycleCopyCatalog.undoStartExpired(month: month))
            case .undoCompletion:
                return canUndoCompletion
                    ? .allowedDecision
                    : blocked(.undoCompletionExpired, MonthlyCycleCopyCatalog.undoCompletionExpired(month: month))
            }

        case .conflict:
            return blocked(.recordConflict, MonthlyCycleCopyCatalog.recordConflict())
        }
    }

    private static func blocked(
        _ key: MonthlyCycleBlockedCopyKey,
        _ message: String
    ) -> MonthlyCycleActionDecision {
        MonthlyCycleActionDecision(
            allowed: false,
            blockedCopyKey: key,
            blockedMessage: message
        )
    }
}
